Add `torch._lazy_clone` to create COW tensors

kurtamohler opened the pull request:
Part of #109833

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113397

kurtamohler committed:
Add `torch._lazy_clone` to create COW tensors

[ghstack-poisoned]

kurtamohler committed:
Update on "Add `torch._lazy_clone` to create COW tensors"




[ghstack-poisoned]

kurtamohler committed:
Update on "Add `torch._lazy_clone` to create COW tensors"




[ghstack-poisoned]

kurtamohler committed:
Update on "Add `torch._lazy_clone` to create COW tensors"




[ghstack-poisoned]

kurtamohler committed:
Update on "Add `torch._lazy_clone` to create COW tensors"




[ghstack-poisoned]

kurtamohler committed:
Update on "Add `torch._lazy_clone` to create COW tensors"




[ghstack-poisoned]

kurtamohler committed:
Update on "Add `torch._lazy_clone` to create COW tensors"




[ghstack-poisoned]

kurtamohler committed:
Update on "Add `torch._lazy_clone` to create COW tensors"




[ghstack-poisoned]

kurtamohler committed:
Update on "Add `torch._lazy_clone` to create COW tensors"




[ghstack-poisoned]

kurtamohler commented:
The failing job [pull / linux-focal-py3_8-clang9-xla / test (xla, 1, 1, linux.12xlarge)](https://hud.pytorch.org/pr/pytorch/pytorch/113397#18578646970) was marked as flaky, but it actually does have a genuine failure

kurtamohler commented:
I'm not exactly sure what's happening with XLA. The problem is that when `torch._lazy_clone(t)` is called on an XLA tensor, it apparently does not replace the DataPtr of the original tensor, since `self.assertTrue(torch._C._is_cow_tensor(t))` is failing in CI.

I ssh-ed into the CI job to try to debug the problem. First I ran this gdb command that just creates an XLA tensor and tries to create a lazy clone of it:

```
gdb --args python -c 'import torch; import torch_xla; import torch_xla.core.xla_model as xm; a = torch.randn(2, 2, device=xm.xla_device()); a_clone = a._lazy_clone()'
```

Debugging this with gdb did not go very well. I was able to create a breakpoint in the `c10::impl::cow::lazy_clone_storage` by running `b _ZN3c104impl3cow18lazy_clone_storageERNS_11StorageImplE` (found with `nm -A workspace/build/lib/libc10.so | grep 'lazy_clone'`). That breakpoint works, and I can get a backtrace there:

<details><summary>Click to expand</summary>

```
Thread 1 "python" hit Breakpoint 1, 0x00007fffe99c0dd0 in c10::impl::cow::lazy_clone_storage(c10::StorageImpl&) () from /opt/conda/lib/python3.8/site-packages/torch/lib/libc10.so
(gdb) bt
#0  0x00007fffe99c0dd0 in c10::impl::cow::lazy_clone_storage(c10::StorageImpl&) ()
   from /opt/conda/lib/python3.8/site-packages/torch/lib/libc10.so
#1  0x00007fffeabf0abd in at::native::_lazy_clone(at::Tensor const&) ()
   from /opt/conda/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#2  0x00007fffec0f4400 in c10::impl::make_boxed_from_unboxed_functor<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor (at::Tensor const&), &at::(anonymous namespace)::(anonymous namespace)::wrapper_CompositeExplicitAutograd___lazy_clone>, at::Tensor, c10::guts::typelist::typelist<at::Tensor const&> >, false>::call(c10::OperatorKernel*, c10::OperatorHandle const&, c10::DispatchKeySet, std::vector<c10::IValue, std::allocator<c10::IValue> >*) () from /opt/conda/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#3  0x00007fffea7ef64a in c10::Dispatcher::callBoxed(c10::OperatorHandle const&, std::vector<c10::IValue, std::allocator<c10::IValue> >*) const ()
   from /opt/conda/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#4  0x00007fffea7ea862 in void c10::BoxedKernel::make_boxed_function<&(anonymous namespace)::functionalizeFallback>(c10::OperatorKernel*, c10::OperatorHandle const&, c10::DispatchKeySet, std::vector<c10::IValue, std::allocator<c10::IValue> >*) ()
   from /opt/conda/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#5  0x00007fffebb008e2 in at::_ops::_lazy_clone::redispatch(c10::DispatchKeySet, at::Tensor const&) () from /opt/conda/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#6  0x00007fffedc73fc2 in c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor (c10::DispatchKeySet, at::Tensor co--Type <RET> for more, q to quit, c to continue without paging--
nst&), &torch::autograd::VariableType::(anonymous namespace)::_lazy_clone>, at::Tensor, c10::guts::typelist::typelist<c10::DispatchKeySet, at::Tensor const&> >, at::Tensor (c10::DispatchKeySet, at::Tensor const&)>::call(c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&)
    () from /opt/conda/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#7  0x00007fffeba3d508 in at::_ops::_lazy_clone::call(at::Tensor const&) ()
   from /opt/conda/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#8  0x00007ffff61649f1 in torch::autograd::THPVariable__lazy_clone(_object*, _object*) ()
   from /opt/conda/lib/python3.8/site-packages/torch/lib/libtorch_python.so
...
```

</details>

However, I can't get `list` to show me the source:

```
(gdb) list
2	/usr/local/src/conda/python-3.8.18/Programs/python.c: No such file or directory.
```

I can look at the assembly of course, but for some reason when I try to step just one instruction with `step`, it runs until the end of the function call, so I'm having trouble seeing what's going on.

I also tried creating breakpoints for the `has_simple_data_ptr` and `is_cow_data_ptr` functions, to try to see which of the three cases get chosen in `lazy_clone_storage`. No luck there though--those two functions must have been inlined because the breakpoints I set for them don't trigger.

I started just trying to get more information about what's happening by running straight Python, and I found this:

```
>>> import torch
>>> import torch_xla
>>> import torch_xla.core.xla_model as xm
>>> a = torch.randn(2, 2, device=xm.xla_device())
>>> torch._C._storage_address(a)
107292784
>>> torch._C._data_address(a)
0
```

The data pointer of XLA tensors is 0. Knowing that, I now would not expect `lazy_clone_storage` to work correctly for XLA, since it depends on checking the value of the data pointer.

Just to double check, I verified that `_lazy_clone` does not create COW tensors:

```
>>> a_clone = torch._lazy_clone(a)   
>>> torch._C._is_cow_tensor(a_clone)
False
>>> torch._C._is_cow_tensor(a)
False
```

Maybe for now it makes sense to not support XLA tensors in `torch._lazy_clone`, but I would like to understand how they work.

kurtamohler commented:
Would it make sense to make `_lazy_clone` just call `clone` for XLA? Since XLA already runs lazily, it seems like there may not be any need to create COW XLA tensors

ezyang commented:
I haven't read the rest yet, but yes, punt XLA for later

kurtamohler committed:
Update on "Add `torch._lazy_clone` to create COW tensors"

Part of #109833




[ghstack-poisoned]

kurtamohler committed:
Update on "Add `torch._lazy_clone` to create COW tensors"

Part of #109833




[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

kurtamohler commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change but the lower PR in the stack https://github.com/pytorch/pytorch/pull/113396 is failing a bunch of internal build jobs.  I will provide the detail on https://github.com/pytorch/pytorch/pull/113396' -c ghfirst

huydhn commented:
Ouch, revert conflicts.  I will check with the team on Monday then.  Basically, the build failure is:

```
fbcode/caffe2/caffe2/fb/init/huge_pages_allocator.cc:171:8: error: abstract class is marked 'final' [-Werror,-Wabstract-final-class]
[CONTEXT] struct OversizeArenaAllocator final : at::Allocator {
[CONTEXT]        ^
fbcode/caffe2/caffe2/fb/init/huge_pages_allocator.cc:208:31: error: variable type 'caffe2::(anonymous namespace)::OversizeArenaAllocator' is an abstract class
[CONTEXT] static OversizeArenaAllocator g_oversize_cpu_alloc;
[CONTEXT]                               ^
```

So using `virtual void copy_data(void* dest, const void* src, std::size_t count) const = 0;` is harder to roll out without updating the usage of Allocator class internally.

DanilBaibak commented:
@pytorchbot revert -m 'Unfortunately, I need to revert your PR because the lower [PR in the stack](https://github.com/pytorch/pytorch/pull/113396) is failing a bunch of internal build jobs.' -c ghfirst

ezyang commented:
@ezyang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51466281).

kurtamohler committed:
Update on "Add `torch._lazy_clone` to create COW tensors"

Part of #109833


Differential Revision: [D51466281](https://our.internmc.facebook.com/intern/diff/D51466281)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51466281).

kurtamohler committed:
Update on "Add `torch._lazy_clone` to create COW tensors"

Part of #109833


Differential Revision: [D51466281](https://our.internmc.facebook.com/intern/diff/D51466281)

[ghstack-poisoned]

kurtamohler committed:
Update on "Add `torch._lazy_clone` to create COW tensors"

Part of #109833


Differential Revision: [D51466281](https://our.internmc.facebook.com/intern/diff/D51466281)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51466281).

kurtamohler commented:
The failure in [cuda12.1-py3.10-gcc9-sm86 / test (inductor, 1, 1, linux.g5.4xlarge.nvidia.gpu)](https://github.com/pytorch/pytorch/actions/runs/6964405236/job/18952263854#logs) is real. Running `test_lazy_clone` with `PYTORCH_TEST_WITH_INDUCTOR=1` does not work, but apparently it can fail silently for some reason.

I just found #114012, which skips the inductor test. I'll add that change to this PR instead.

I'm not sure why the inductor tests are failing. I think it probably makes sense to fix it in a follow up

kurtamohler committed:
Update on "Add `torch._lazy_clone` to create COW tensors"

Part of #109833


Differential Revision: [D51466281](https://our.internmc.facebook.com/intern/diff/D51466281)

[ghstack-poisoned]

kurtamohler committed:
Update on "Add `torch._lazy_clone` to create COW tensors"

Part of #109833


Differential Revision: [D51466281](https://our.internmc.facebook.com/intern/diff/D51466281)

[ghstack-poisoned]

kurtamohler commented on test/test_torch.py:
```diff
@@ -5042,6 +5042,130 @@ def test_storage_all_devices(self, devices):
             t = torch.tensor((), device=device)
             self.assertEqual(t.dtype, t.storage().dtype)
 
+    # Note [lazy_clone_ tests with inductor enabled]
```
I was able to split the test up into multiple tests to get them to pass when inductor is enabled. This note explains what went wrong and the strategy I used to fix it

vadimkantorov commented:
Related idea on copy-on-write/move-semantics: https://github.com/pytorch/pytorch/issues/106584#issuecomment-1685398317

And expressing that clone is not always needed for currently always-copy functions, such as torch.cat with a single-tensor argument

ezyang committed:
Update on "Add `torch._lazy_clone` to create COW tensors"

Part of #109833


Differential Revision: [D51466281](https://our.internmc.facebook.com/intern/diff/D51466281)

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge -f "trunk flakiness only"
